plugins/outputs/postgresql/postgresql.go

```diff
@@ -0,0 +1,424 @@
+	TagTableCreateTemplates    []*template.Template
+	TagTableAddColumnTemplates []*template.Template
+	RetryMaxBackoff            config.Duration
+	LogLevel                   string
+
+	dbContext       context.Context
+	dbContextCancel func()
+	db              *pgxpool.Pool
+	tableManager    *TableManager
+	tagsCache       *freecache.Cache
+
+	writeChan      chan *TableSource
+	writeWaitGroup *utils.WaitGroup
+
+	Logger telegraf.Logger
+}
+
+func init() {
+	outputs.Add("postgresql", func() telegraf.Output { return newPostgresql() })
+}
+
+func newPostgresql() *Postgresql {
+	p := &Postgresql{
+		Logger: models.NewLogger("outputs", "postgresql", ""),
+	}
+	if err := toml.Unmarshal([]byte(p.SampleConfig()), p); err != nil {
+		panic(err.Error())
+	}
+	return p
+}
+
+func (p *Postgresql) SampleConfig() string { return sampleConfig }
+func (p *Postgresql) Description() string  { return "Send metrics to PostgreSQL" }
+
+// Connect establishes a connection to the target database and prepares the cache
+func (p *Postgresql) Connect() error {
+	poolConfig, err := pgxpool.ParseConfig(p.Connection)
+	if err != nil {
+		return err
+	}
+	parsedConfig, _ := pgx.ParseConfig(p.Connection)
+	if _, ok := parsedConfig.Config.RuntimeParams["pool_max_conns"]; !ok {
+		// The pgx default for pool_max_conns is 4. However we want to default to 1.
+		poolConfig.MaxConns = 1
+	}
+
+	if p.LogLevel != "" {
+		poolConfig.ConnConfig.Logger = utils.PGXLogger{p.Logger}
+		poolConfig.ConnConfig.LogLevel, err = pgx.LogLevelFromString(p.LogLevel)
+		if err != nil {
+			return fmt.Errorf("invalid log level")
+		}
+	}
+
+	// Yes, we're not supposed to store the context. However since we don't receive a context, we have to.
+	p.dbContext, p.dbContextCancel = context.WithCancel(context.Background())
+	p.db, err = pgxpool.ConnectConfig(p.dbContext, poolConfig)
+	if err != nil {
+		p.Logger.Errorf("Couldn't connect to server\n%v", err)
+		return err
+	}
+	p.tableManager = NewTableManager(p)
+
+	if p.TagsAsForeignKeys {
+		p.tagsCache = freecache.NewCache(5 * 1024 * 1024) // 5MB
+	}
+
+	maxConns := int(p.db.Stat().MaxConns())
+	if maxConns > 1 {
+		p.writeChan = make(chan *TableSource)
+		p.writeWaitGroup = utils.NewWaitGroup()
+		for i := 0; i < maxConns; i++ {
+			p.writeWaitGroup.Add(1)
+			go p.writeWorker(p.dbContext)
+		}
+	}
+
+	return nil
+}
+
+// Close closes the connection(s) to the database.
+func (p *Postgresql) Close() error {
+	if p.writeChan != nil {
+		// We're using async mode. Gracefully close with timeout.
+		close(p.writeChan)
+		select {
+		case <-p.writeWaitGroup.C():
+		case <-time.NewTimer(time.Second * 5).C:
+		}
+	}
+
+	// Die!
+	p.dbContextCancel()
+	p.db.Close()
+	p.tableManager = nil
+	return nil
+}
+
+func (p *Postgresql) Write(metrics []telegraf.Metric) error {
+	tableSources := NewTableSources(p, metrics)
+
+	if p.db.Stat().MaxConns() > 1 {
+		return p.writeConcurrent(tableSources)
+	} else {
+		return p.writeSequential(tableSources)
+	}
```

Comment:
Please use
```suggestion
	}
	return p.writeSequential(tableSources)
```
to fix the linter, as you can only reach this code if not entering the `if`.